### PR TITLE
loader: attach programs using tcx

### LINKED ISSUE #27632: Attach TC programs using bpf_link tcx instead of netlink
Depends on https://github.com/cilium/ebpf/issues/1117.

This issue is for replacing the current tc filter creation and attachment logic with a tcx-based API.

Design TBD since the ebpf-go API hasn't been written, but we could pin links at `/sys/fs/bpf/cilium/{endpoints||tcx}/lxcXYZ/{ingress,egress}`.

### PR DESCRIPTION
**Pull request by rgo3**
For more detailed descriptions, please refer to the individual commits.

On a high level, this PR:

- attaches TC progs using bpf_link (tcx) respecting upgrade and downgrade paths
- uses per-endpoint bpffs dirs, at e.g. /sys/fs/bpf/cilium/endpoints/12345/links/cil_to_container
- adds netns-driven tests for attaching skb progs via tcx
- adds a Helm flag (enableTXC) to optionally disable the feature to ease integration with other tools

Closes https://github.com/cilium/cilium/issues/27632

### CODE REVIEW

File: pkg/datapath/loader/netlink.go
Comment (rgo3) [inline pkg/datapath/loader/netlink.go]: idea: should this query for a specific program name instead?

**Comment by rgo3** [comment]
Connectivity tests fail because currently the TCX API in cilium/ebpf doesn't do a feature test, so on kernels without TCX support creating links fails and we don't fall back to the old behavior. Opened https://github.com/cilium/ebpf/pull/1294 to add a feature test in ebpf-go.

The build every commit action will continue to fail because I've separate commits for vendoring the lib changes and fixing the query programs API breakage. I'll keep them separate so the fix can easily be applied to the renovate PR that updates the ebpf-go dependency.

**Comment by ti-mo** [comment]
Needs rebase on top of https://github.com/cilium/cilium/pull/31553 when merged.

**Comment by rgo3** [comment]
ACK for Timos refactorings. Can't actually give the PR real approval, because it's my own PR :grin:

File: pkg/datapath/loader/netlink.go
Comment (borkmann) [inline pkg/datapath/loader/netlink.go]: Looks good to me! Thanks Robin!

**Review by borkmann — APPROVED** [review]

File: pkg/datapath/loader/netlink.go
Comment (rgo3) [inline pkg/datapath/loader/netlink.go]: Nit: Should this be DeviceHasSKBProgramLoaded for consistency?

File: pkg/datapath/loader/tc_test.go
Comment (rgo3) [inline pkg/datapath/loader/tc_test.go]: Should this test be run with tcx disabled? We already cover all the TCX cases with other tests, so this one should maybe focus on legacy tc attach/detach.

File: pkg/datapath/loader/tc_test.go
Comment (rgo3) [inline pkg/datapath/loader/tc_test.go:136]: This test should be skipped if TCX is not supported, similar to the other tests in tcx_test.go.

File: pkg/datapath/loader/tc_test.go
Comment (rgo3) [inline pkg/datapath/loader/tc_test.go:167]: skippidyskip?

**Review by rgo3 — COMMENTED** [review]
Overall LGTM, just a couple of nits. Wrt docs, should we mention/explain that enabling tcx will skip any other existing legacy tc attachments?

**Comment by borkmann** [comment]
> Overall LGTM, just a couple of nits. Wrt docs, should we mention/explain that enabling tcx will skip any other existing legacy tc attachments?

File: pkg/datapath/loader/netlink.go
Comment (ti-mo) [reply pkg/datapath/loader/netlink.go]: Good idea! Also removed it from interface `Loader`, it didn't actually access the underlying struct..

File: pkg/datapath/loader/tc_test.go
Comment (ti-mo) [reply pkg/datapath/loader/tc_test.go]: Good point, I think I had this in mind but it fell through. I ended up renaming this to `TestAttachDetachSKBProgramLegacy` and adding a separate `TestAttachDetachTCProgram` that only uses the legacy tc APIs. :+1:

File: install/kubernetes/cilium/values.yaml
Comment (gandro) [inline install/kubernetes/cilium/values.yaml]: nit: Would it make sense to put this under the `bpf` map, i.e. make the value `bpf.enableTCX`?

**Review by gandro — APPROVED** [review]
My CODEOWNERs look good to me, thanks!

File: install/kubernetes/cilium/values.yaml
Comment (ti-mo) [reply install/kubernetes/cilium/values.yaml]: That makes sense to me, but to be honest I'm not sure. For example, `bpfClockProbe` also sits in the root, which IMO should be exposed as `bpf.clockProbe`. The whole chart is consistently inconsistent.

tcx is technically bpf-specific, but the legacy tc subsystem it replaces isn't, so I was a bit skeptical. I guess I'll bite the bullet, I don't like it sitting in the root either.

Comment (ti-mo) [reply install/kubernetes/cilium/values.yaml]: @gandro Done, and I simplified the chart logic a bit (always rendering `enable-tcx` to the configmap), PTAL.

**Review by gandro — APPROVED** [review]

File: install/kubernetes/cilium/values.yaml
Comment (gandro) [reply install/kubernetes/cilium/values.yaml]: Looks good, thanks! Yeah, our charts are super inconsistent unfortunately

Comment (borkmann) [reply install/kubernetes/cilium/values.yaml]: At some point probably makes sense to bite the bullet and clean it up for more consistency where feasible. Anyway, moving under bpf makes sense. Lgtm.

**Review by ti-mo — APPROVED** [review]

File: pkg/datapath/loader/loader.go
Comment (jrfastab) [inline pkg/datapath/loader/loader.go:355]: Is this chunk a bugfix? Seems we should have always removed this dir or is this the first link there?

Comment (borkmann) [reply pkg/datapath/loader/loader.go:355]: No, given this is the first time we're using tcx links.

File: pkg/datapath/loader/tcx.go
Comment (jrfastab) [inline pkg/datapath/loader/tcx.go]: this is all a bit odd to me. I would either return from all case statements or none. But just a nitpick.

File: pkg/datapath/loader/tc.go
Comment (jrfastab) [inline pkg/datapath/loader/tc.go]: if we throw an error here will upsertTCXProgram leave state changed? Should we revert change to state when returning an error if this is the case?

**Comment by jrfastab** [comment]
Some minor comments/questions about the code otherwise lgtm.

File: pkg/datapath/loader/tc.go
Comment (borkmann) [reply pkg/datapath/loader/tc.go]: Good point, in this case we could either add a `detachTCX(bpffsDir, progName)` w/o error checking before the return (given we cannot recover from that if the latter fails :/) or we just call the `removeTCFilters()` without checking error given tcx installed successfully and won't interfere with tcx at least.. not sure which is better tbh, maybe the latter - technically none of them should ever fail.

**Review by learnitall — APPROVED** [review]
API changes LGTM, thanks!

**Review by ghost — APPROVED** [review]

File: pkg/datapath/loader/tc.go
Comment (borkmann) [reply pkg/datapath/loader/tc.go]: (I will follow-up on this.)

File: pkg/datapath/loader/tcx.go
Comment (ti-mo) [reply pkg/datapath/loader/tcx.go]: Yeah it does look weird, I'll fix.

File: pkg/datapath/loader/tc.go
Comment (ti-mo) [reply pkg/datapath/loader/tc.go]: Yeah, let's not bail out if tcx already attached successfully and removing the legacy filter failed somehow.

**Review by tklauser — APPROVED** [review]

**Comment by joestringer** [comment]
I wonder if this is something we should elevate to `release-note/major` given the potential performance impacts. This seems like a great thing to tout in the release blog.

**Comment by julianwiedmann** [comment]
:wave: looking at https://github.com/cilium/cilium/actions/runs/8931592322. What's the current upgrade/downgrade impact of this feature - in particular without https://github.com/cilium/cilium/pull/32228 merged? Do we need to switch some of the `bpf-next` downgrade CI tests to `bpf.enableTCX=false`, until v1.15 is sorted?

**Comment by borkmann** [comment]
> 👋 looking at https://github.com/cilium/cilium/actions/runs/8931592322. What's the current upgrade/downgrade impact of this feature - in particular without #32228 backported? Do we need to switch some of the `bpf-next` downgrade CI tests to `bpf.enableTCX=false`, until v1.15 is sorted?

I just did the backport of mentioned PR in https://github.com/cilium/cilium/pull/32337 . The ci-ipsec-upgrade test I've seen sporadically fail on various kernels though, not only bpf-next.

**Comment by borkmann** [comment]
> What's the current upgrade/downgrade impact of this feature - in particular without https://github.com/cilium/cilium/pull/32228 backported?

Without the backport there is a tiny race for downgrade where tcx gets removed and legacy tc installed so for a short window there is nothing attached. With the backport it's atomic. The upgrade is also atomic, that is, tcx is installed and then tc legacy removed (if tcx is installed successfully it takes precedence and legacy tc is not run). Lets see if the backport helps.

**Comment by julianwiedmann** [comment]
> > What's the current upgrade/downgrade impact of this feature - in particular without #32228 backported?
> 
> Without the backport there is a tiny race for downgrade where tcx gets removed and legacy tc installed so for a short window there is nothing attached. With the backport it's atomic. The upgrade is also atomic, that is, tcx is installed and then tc legacy removed (if tcx is installed successfully it takes precedence and legacy tc is not run). Lets see if the backport helps.

I'm at attempt 7 for the `bpf-next` downgrade from main to v1.15 in https://github.com/cilium/cilium/actions/runs/8927670014, which includes the txc work. And as the workflow is *meant* to be sensitive enough for such tiny blips, I'm very inclined to blame this race window ;).

Can we please flip the default to `bpf.enableTCX=false`, until the next v1.15 patch release is through and we can offer a smooth downgrade?

**Comment by borkmann** [comment]
> > > What's the current upgrade/downgrade impact of this feature - in particular without #32228 backported?
> > 
> > 
> > Without the backport there is a tiny race for downgrade where tcx gets removed and legacy tc installed so for a short window there is nothing attached. With the backport it's atomic. The upgrade is also atomic, that is, tcx is installed and then tc legacy removed (if tcx is installed successfully it takes precedence and legacy tc is not run). Lets see if the backport helps.
> 
> I'm at attempt 7 for the `bpf-next` downgrade from `main` to `v1.15` in https://github.com/cilium/cilium/actions/runs/8927670014, which includes the txc work. And as the workflow is _meant_ to be sensitive enough for such tiny blips, I'm very inclined to blame this race window ;).
> 
> Can we please flip the default to `bpf.enableTCX=false`, until the next v1.15 patch release is through and we can offer a smooth downgrade?

Ah yes, I thought the downgrade tests would test against latest v1.15 branch, not patch release. Cc @ti-mo lets default to false until then wdyt?

**Comment by borkmann** [comment]
(PR is here https://github.com/cilium/cilium/pull/32342 )

**Comment by rgo3** [comment]
@julianwiedmann I've retriggered the job you referenced previously and it is now passing: https://github.com/cilium/cilium/actions/runs/8927670014/job/24559664765. Your 7 runs prior to that pulled a `v1.15` image before we merged the seamless downgrade path 4h ago. So I don't think `bpf.enableTCX=false` is necessary anymore.

**Comment by julianwiedmann** [comment]
> @julianwiedmann I've retriggered the job you referenced previously and it is now passing: https://github.com/cilium/cilium/actions/runs/8927670014/job/24559664765. Your 7 runs prior to that pulled a `v1.15` image before we merged the seamless downgrade path 4h ago. So I don't think `bpf.enableTCX=false` is necessary anymore. I'll also rerun it a couple times more to gather some additional data points.

:shrug: I give up on understanding the downgrade tests. So the **minor** downgrade *is* running against the previous minor's branch tip, and **not** the previous minor's current release. And from a quick git archaeology, that was always the case? Then we should be good indeed. Sorry for the noise. Thank you @rgo3 for keeping me honest!

**Comment by rgo3** [comment]
I feel the same :sweat_smile:. 

But yes, I believe and have seen in the CI logs that we install previous minor's branch tip, in this case `v1.15`. To me that also makes the most sense for the minor tests, given the previous minor's branch tip will be the code we support minor downgrades to once released. Otherwise, we would have to backport things, wait until it's merged and released, and only then can move forward with the PR for `main` (something we did in the past, including this PR), which always felt quite tedious to me.

**Pull request merged**
